lib/grandpa/finalization.go

```diff
@@ -0,0 +1,493 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/telemetry"
+)
+
+type ephemeralService interface {
+	Start() error
+	Stop() error
+}
+
+type finalizationHandler struct {
+	servicesLock       sync.Mutex
+	finalizationEngine ephemeralService
+	votingRound        ephemeralService
+
+	newServices    func() (engine, voting ephemeralService)
+	timeoutStop    time.Duration
+	initiateRound  func() error
+	observableErrs chan error
+	stopCh         chan struct{}
+	handlerDone    chan struct{}
+}
+
+func newFinalizationHandler(service *Service) *finalizationHandler {
+	// builder is a constructor function which takes care to instantiate
+	// and return the services needed to finalize a round, those services
+	// are ephemeral services with a lifetime of a round
+	builder := func() (engine, voting ephemeralService) {
+		finalizationEngine := newFinalizationEngine(service)
+		votingRound := newHandleVotingRound(service, finalizationEngine.actionCh)
+		return finalizationEngine, votingRound
+	}
+
+	return &finalizationHandler{
+		newServices: builder,
+
+		timeoutStop:    5 * time.Second,
+		initiateRound:  service.initiateRound,
+		observableErrs: make(chan error),
+		stopCh:         make(chan struct{}),
+		handlerDone:    make(chan struct{}),
+	}
+}
+
+func (fh *finalizationHandler) Start() (errsCh <-chan error, err error) {
+	go fh.start()
+	return fh.observableErrs, nil
+}
+
+func (fh *finalizationHandler) start() {
+	defer close(fh.handlerDone)
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return
+		default:
+		}
```

Comment:
`waitServices` already returns when `stopCh` is closed since it should shut down the inner services it manages.

`initiateRound` does not have to return when the `stopCh` is closed because it does not manage the inner services, it basically updates the authorities (if needed) and increases the round number, so once this is done this function returns.

Reply:
`initiateRound` does look it doesn't do IO/infinite looping, my bad. Although it does lock mutexes and stuff, so let's just assume/hope it doesn't deadlock on exit.

Regarding `waitServices`, maybe we could inject the stop channel as an argument `waitServices(fh.stopCh)` from the caller? That way we can also restrict it as read only channel. I guess I'm used to passing contexts as arguments and not as field, and this feels like a context-ish.

Reply:
Since the `stopCh` is a field and `waitServices ` is a method of the `finalizationHandler` I don't see why we should pass `stopCh` to `waitServices ` as an argument given that we could just use it from the instance

Reply:
There's a lot to unpack here.  I don't understand what you mean by "we stop both ephemeral services kinda at the wrong place (might be causing one or more problems as well)".  If you're going to make a declaration like this you need to elaborate why your reasoning behind this.

> It really does help to inject all channels as arguments (stop channel, error channel, ready channel etc.), even if it's already present in the receiving struct. We really just need to have them in the receiving struct so both Start and Stop can share some of the channels, but then internally, we should inject them as arguments for our own sakes.

Again, what does this mean? How does it help? We need the stop chan to be in the state of the service because they need to be read by one receiver method and accessed by another.  How does passing it via parameters help this? We would need to introduce new methods that have different signatures than the methods that are publicly exposed.  

> I'm pausing my local suggestion rework since we should first throw out the service nature of votingRound (just make it a single global function handleVotingRound(service *Service, action engineAction) without even a votingRound struct).

@EclesioMeloJunior designed `handleVotingRound` to be asynchronous for a reason.  I think it mirrors more of how the `sc-grandpa-finality` interacts with `grandpa-finality` package in substrate.  I don't see why you wouldn't want to process more incoming votes or determine whether or not we can continue to a completable state if we're waiting on a synchronous call to said function.

Reply:
> I don't understand what you mean by "we stop both ephemeral services kinda at the wrong place (might be causing one or more problems as well)". If you're going to make a declaration like this you need to elaborate your reasoning behind this.

If the finalization engine fails, the voting round handler is then stopped in `runEphemeralServices() error` (same the other way around). If the caller then calls `Stop()` on the finalization handler, it would close signal channels twice and panic. I think Eclesio and I went over this Friday afternoon. The solution would be to stop both services within `runEphemeralServices() error` (and nowhere else) in the `case <-fh.stopCh:`. It gets a bit more complicated though, we would have to write the error from stopping both services to an extra internal channel so it can be retrieved from the `Stop()` method.

> Again, what does this mean? How does it help?

It's easier to track what function gets what, especially since channels/parallel programming isn't particularly easy. Just my opinion/own experience though, take it as you wish. 

> We need the stop chan to be in the state of the service because they need to be read by one receiver method and accessed by another. How does passing it via parameters help this? We would need to introduce new methods that have different signatures than the methods that are publicly exposed.

That's what I said, and I didn't suggest to change the exported method signatures: *We really just need to have them in the receiving struct so both Start and Stop can share some of the channels but then internally, we should inject them as arguments for our own sakes.* I meant to pass down the channels in the call stack below `Start`/`Stop` in case that wasn't clear.

> @EclesioMeloJunior designed handleVotingRound to be asynchronous for a reason. I think it mirrors more of how the sc-grandpa-finality interacts with grandpa-finality package in substrate. I don't see why you wouldn't want to process more incoming votes or determine whether or not we can continue to a completable state if we're waiting on a synchronous call to said function.

I just saw the current state of it when reviewing the code, where two goroutines with non-cancelable runs (within their infinite for loops) were communicating through an unbuffered channel, and that's pretty much a function call with extra unneeded cpu locking. The channel would be buffered, then it would make sense. I then learnt Eclesio wants to have some sort of queuing between the two goroutines, so it makes sense to have it like that then (although hiding the queuing within the voting round handler).

Reply:
My question what is the expected behavior on the caller here? If one of the ephemeral services stops, you say then if caller calls Stop() it will panic, not good, but will this happen? It feels to me like its a binary thing, either this is the intended behavior or not. If it is, then we need to stop it from panicking, yes? But if its not then its not something we need to worry about (we shouldnt/wont just be arbitrarily calling stop on our finality service I wouldn't assume)

Reply:
In our code this cannot happen. Right now we use this internally and it's not even exported (`finalisationHandler`). I'm thinking probably way too ahead to have a usable public Go API. BUT I believe :pray: :angel: in ourselves one day we'll expose all this stuff to the outside Go world :tada: And yes we should strike a balance between the two, and if this is to be further refactored there can be no point at all making it better as well.

Coming back to that particular problem and expected behavior: from an Go user perspective, I think it would easily make sense to call `.Stop` (even if it does nothing) if a runtime error is read from a returned channel. On the other hand, this doesn't extend to, in example, calling `.Stop` twice.

Anyway I guess we won't do anything further in this PR :smile: But it's good to discuss and exchange opinions :+1: